### ISSUE TITLE
Spurious dependencies make upgrading packages painful

### ISSUE THREAD

**Issue by larsyencken** [owid/etl#696]
## Problem

Suppose we need to update a package dependency like `pyarrow`. Currently it's a nightmare to do, given our various repos and their interdependencies.

## Current package dependencies

Today we have a dependency structure that looks like:

```mermaid
graph LR

etl --> walden --> owid-datautils-py
walden --> owid-catalog-py
etl --> owid-catalog-py
etl --> owid-datautils-py --> owid-catalog-py
```

The problem is that it's very hard to update a package requirement like `pyarrow` across all of these dependencies. It takes tons of pull requests, and at least one PyPI release.

## Ideal dependencies

Ideally, our packages would be more independent and far less interlinked:

```mermaid
graph LR
etl --> walden
etl --> owid-catalog-py
etl --> owid-datautils-py
```

- `walden --> owid-datautils-py`: only uses a single helper for serialising to disk => just copy it into walden
- `walden --> owid-catalog-py`: uses frame repacking => make a separate frame-packing lib with generous dependency range
- `owid-datautils-py --> owid-catalog-py`: it's pulling population data, which can cause correctness issues and subverts the ETL's dependency graph; we want to remove this link entirely

If we split out frame repacking, it would look more like this:

```mermaid
graph LR
etl --> walden --> owid-repack-py
etl --> owid-catalog-py --> owid-repack-py
etl --> owid-datautils-py --> owid-repack-py
```

## Monorepo preferable?

Alternatively, we could consider a monorepo layout to try to make it possible to update all these libs in one commit. An example layout for this borrowed from OpenDoor is: https://github.com/owid/owid-mono-py

The monorepo layout is much nicer than vendoring, which is what we're doing for the ETL. It still requires multiple PRs to update public libraries like `owid-datautils-py`, but it's a better environment for doing that.

**Comment by larsyencken** [comment]
Heads up @pabloarosado @lucasrodes @Marigold 

This stuff tripped me up today. I avoided doing a push-to-master frenzy, barely, but I may do it tomorrow just to fix the `pyarrow` issues and get Python 3.11 working 😅

**Comment by Marigold** [comment]
I'm up for giving monorepo a try. Moving to monorepo has never been a bad thing from my experience.

We could also wait until I migrate Walden to Snapshots, but that's blocked by datasets diff in CI (which is WIP). That said... go berserk Lars!!!

**Comment by lucasrodes** [comment]
Thanks for raising this, @larsyencken.

FYI, `owid-datautils-py` should soon not have any dependency on `owid-catalog-py`. See https://github.com/owid/etl/pull/682 for more context.

While mono-repo could ease some of our work, I think it might add some complexity to the repo (plus lots of commits), which might make it hard to follow for newcomers (unless beautiful docs are there, which is also an option).

If my understanding is correct, with the migration from `walden` to `snapshots` and dependency removal of `datautils` from `catalog`, we will soon have a dependency structure like:

```mermaid
graph LR

etl --> owid-catalog-py
etl --> owid-datautils-py
```

Not sure I understand the purpose of `owid-repack-py` though. What would it's usage be?

**Comment by larsyencken** [comment]
Data frame repacking is a useful, standalone thing, that's totally independent of the rest of our work. It's been a candidate to break out for ages. It's now its own repo: https://github.com/owid/owid-repack-py

The new state of affairs is this:

```mermaid
graph LR

etl --> owid-catalog --> owid-repack
etl --> owid-repack
etl --> walden
walden --> owid-repack
walden --> owid-datautils
etl --> owid-datautils -.-> owid-catalog
```
